floris/tools/optimization/layout_optimization/layout_optimization_scipy.py

```diff
@@ -0,0 +1,316 @@
+            self._unnorm(valx, self.bndx_min, self.bndx_max)
+            for valx in x_in[0 : self.nturbs]
+        ] 
+        y =  [
+            self._unnorm(valy, self.bndy_min, self.bndy_max)
+            for valy in x_in[self.nturbs : 2 * self.nturbs]
+        ]
+        # x = np.array(x_in[0 : self.nturbs])
+        # y = np.array(x_in[self.nturbs :])
+        # Calculate distances between turbines
+        locs = np.vstack((x, y)).T
+        distances = cdist(locs, locs)
+        arange = np.arange(distances.shape[0])
+        distances[arange, arange] = 1e10
+        dist = np.min(distances, axis=0)
+
+        g = 1 - np.array(dist) / self.min_dist
+
+        # Following code copied from OpenMDAO KSComp().
+        # Constraint is satisfied when KS_constraint <= 0
+        g_max = np.max(np.atleast_2d(g), axis=-1)[:, np.newaxis]
+        g_diff = g - g_max
+        exponents = np.exp(rho * g_diff)
+        summation = np.sum(exponents, axis=-1)[:, np.newaxis]
+        KS_constraint = g_max + 1.0 / rho * np.log(summation)
+
+        return -1*KS_constraint[0][0]
+
+    def _distance_from_boundaries(self, x_in):
+        # x = np.array(x_in[0 : self.nturbs])
+        # y = np.array(x_in[self.nturbs :])
+        x = [
+            self._unnorm(valx, self.bndx_min, self.bndx_max)
+            for valx in x_in[0 : self.nturbs]
+        ] 
+        y =  [
+            self._unnorm(valy, self.bndy_min, self.bndy_max)
+            for valy in x_in[self.nturbs : 2 * self.nturbs]
+        ]
+        boundary_con = np.zeros(self.nturbs)
+        for i in range(self.nturbs):
+            loc = Point(x[i], y[i])
+            boundary_con[i] = loc.distance(self.boundary_line)
+            if self.boundary_polygon.contains(loc)==True:
+                boundary_con[i] *= 1.0
+
+        return boundary_con
+
+    # Public methods
+
+    def optimize(self):
+        """
+        This method finds the optimized layout of wind turbines for power
+        production given the provided frequencies of occurance of wind
+        conditions (wind speed, direction).
+
+        Returns:
+            opt_locs (iterable): A list of the optimized locations of each
+            turbine (m).
+        """
+        print("=====================================================")
+        print("Optimizing turbine layout...")
+        print("Number of parameters to optimize = ", len(self.x0))
+        print("=====================================================")
+
+        opt_locs_norm = self._optimize()
+
+        print("Optimization complete.")
+
+        opt_locs = [
+            [
+                self._unnorm(valx, self.bndx_min, self.bndx_max)
+                for valx in opt_locs_norm[0 : self.nturbs]
+            ],
+            [
+                self._unnorm(valy, self.bndy_min, self.bndy_max)
+                for valy in opt_locs_norm[self.nturbs : 2 * self.nturbs]
+            ],
+        ]
+
+        return opt_locs
+
+    def reinitialize_opt(
```

Comment:
Not sure if a `reinitialize_opt` function is necessary. Users can just reinitialize the entire class.

Reply:
Agreed, it was more a function from v2, but with the new mantra in v3 of being more explicit, I think a full reinitialization is correct.